### PR TITLE
Adding subcommands for scripts

### PR DESCRIPTION
**Pull request by kiwiroy**
## Summary

One script, multiple subcommands. Think `git clone`, `git checkout`, etc...

## Motivation

Quite often I wish I could do some similar tasks with one script and achieve code simplicity and sharing. Often these similar tasks have required options which are distinct and force writing two different scripts. This appears, to me at least, to be a solution to this.

I'm happy to expand tests or explain more if there's something missing here.

### CODE REVIEW

**Comment by jhthorsen** [comment]
I like the idea, but I don't like the implementation. I don't want this:

```
app {
  my $self = shift;
  my $subcommand = $self->_script->subcommand;
  if ($subcommand eq 'create') {
    ## do creating
  } else {
    ## do listing
  }
  return 0;
};
```

What I want is rather that we create automatic dispatching to the correct subcommand. Something like this:

```
# if ($subcommand eq 'create')
sub command_create {
  my ($self, @extra_argv) = @_;
  return 0;
}

# } else {
app {
  my ($self, @extra_argv) = @_;
  return 0;
};
```

We could also do something like the code below, but I'm not sure how readable that is:

```
subcommand list => "provide a listing objects" => sub {
  option flag => long      => "long listing";
  option flag => recursive => "recursively list objects";

  app {
    my ($self, @extra_argv) = @_;

    return 0;
  }
};
```

**Comment by kiwiroy** [comment]
You're correct the original option is not elegant, introduces sources of error and complicates things needlessly. I preferred your first suggestion and originally wanted to remove the possibility of the second (there's some [commented code](https://github.com/jhthorsen/applify/pull/4/files#diff-ae417fcb1a78e526fe1931bf85b03ce7R205) to that effect) as readability suffers IMO.

I've made the naming of the *subcommand* functions configurable with `$Applify::SUBCMD_PREFIX` and in doing so considered the option of an alternate name for this feature, along the lines of *action*, *operation* or *task*. I kept it as it is for now.

File: lib/Applify.pm [hunk 1]
```diff
@@ -247,7 +281,7 @@ sub _generate_application_class {
     package $application_class;
     use base qw(@$extends);
     1;
-  ] or die "Failed to generate applicatin class: $@";
+  ] or die "Failed to generate application class: $@";
```
Comment (jhthorsen) [inline lib/Applify.pm]: 👍

File: lib/Applify.pm [hunk 2]
```diff
@@ -344,6 +383,26 @@ sub _sub {
...
+    #   # subcommand 'command' => 'desc' => sub { ... }; called,
+    #   # but no matching command in @ARGV, will display help
+    #   $code = sub {
+    #     my ($self, @extra) = @_;
+    #     $self->_script->print_help;
+    #     return 0;
+    #   };
+    # }
+  }
+  return $code;
```
Comment (jhthorsen) [inline lib/Applify.pm]: Here I would the parsing of @ARGV:

```
sub _subcommand_code {
  my ($self, $name) = @_;
  return undef unless $name and $name =~ /^\w+/;
  return undef unless $self->{subcommands}{$name};
  return $self->can("${SUBCMD_PREFIX}_${name}");
}
```

File: cpanfile [hunk 3]
```diff
@@ -1,6 +1,10 @@
-# You can install this projct with curl -L http://cpanmin.us | perl - https://github.com/jhthorsen/applify/archive/master.tar.gz
+# You can install this project with curl -L http://cpanmin.us | perl - https://github.com/jhthorsen/applify/archive/master.tar.gz
 requires "perl" => "5.010001";
 requires "Cwd" => "3.31",
 requires "File::Basename" => "2.70",
 requires "Getopt::Long" => "2.38",
 test_requires "Test::More" => "0.88";
+
+on develop => sub {
+  requires 'Devel::Cover::Report::Coveralls' => '0.11';
```
[the remark this answers is not in the document]
Comment (kiwiroy) [reply cpanfile]: oops wrong branch.

File: lib/Applify.pm [hunk 4]
```diff
@@ -149,6 +151,19 @@ OPTION:
 
   print "Usage:\n";
 
+  if (exists $self->{subcommands}){
+    my $subcmds = $self->{subcommands};
+    my $width = 0;
+    for my $subcmd(@$subcmds) {
```
[the remark this answers is not in the document]
Comment (kiwiroy) [reply lib/Applify.pm]: switched to a hash for clarity and use `->_has_subcommands()`

File: lib/Applify.pm [hunk 5]
```diff
@@ -17,6 +18,9 @@ sub app {
   my $parser = $self->_option_parser;
   my (%options, @options_spec, $application_class, $app);
 
+  # has to be run before calculating option spec.
+  # cannot do ->can() as application_class isn't created yet.
+  if ($self->_subcommand_activate($ARGV[0])) { shift @ARGV; }
```
Comment (kiwiroy) [inline lib/Applify.pm]: All parsing of `@ARGV` is inside the call to `app` now.

File: lib/Applify.pm [hunk 6]
```diff
@@ -149,6 +154,19 @@ OPTION:
 
   print "Usage:\n";
 
+  if ($self->_has_subcommands){
```
Comment (jhthorsen) [inline lib/Applify.pm]: Just do `%{$self->{subcommands}}` instead.

File: lib/Applify.pm [hunk 7]
```diff
@@ -149,6 +154,19 @@ OPTION:
...
+    my $subcmds = [ sort { $a->{name} cmp $b->{name} } values %{$self->{subcommands}} ];
+    my $width = 0;
+    for my $subcmd(@$subcmds) {
+      my $length = length $subcmd->{name};
+      $width = $length if $width < $length;
+    }
+    print "\n    ", File::Basename::basename($0), " [command] [options]\n";
+    print "\ncommands:\n";
+    printf("    %-${width}s  %s\n", @{$_}{'name', 'desc'}) for @$subcmds;
```
Comment (jhthorsen) [inline lib/Applify.pm]: This  code that use `$subcmds` looks a bit clunky, but I can't come up with anything better myself.

File: lib/Applify.pm [hunk 8]
```diff
@@ -344,6 +375,28 @@ sub _sub {
...
+  return undef unless $self->{subcommands}{$name};
+  $self->{subcommand} = $name;
+  {
+    no warnings 'redefine';
+    local *Applify::app = sub {
+      warn "Avoided deep recursion. Do not call app {} from subcommand block!\n";
+    };
+    $self->{subcommands}{$name}{code}->($self);
+  }
+  $self->{subcommands}{$name}{active} = 1;
```
Comment (jhthorsen) [inline lib/Applify.pm]: I don't think this is used anywhere else. (Remove it)

File: lib/Applify.pm [hunk 9]
```diff
@@ -176,6 +194,13 @@ sub print_version {
   printf "%s version %s\n", File::Basename::basename($0), $version;
 }
 
+sub subcommand {
+  my ($self, $name) = (shift, shift);
+  return $self->{subcommand} unless @_;
+  $self->{subcommands}{$name} = { name => $name, desc => $_[0], code => $_[1] };
```
Comment (jhthorsen) [inline lib/Applify.pm]: I think "code" should be "options" or something.

File: lib/Applify.pm [hunk 10]
```diff
@@ -344,6 +375,28 @@ sub _sub {
...
+sub _subcommand_activate {
+  my ($self, $name) = @_;
+  return undef unless $name and $name =~ /^\w+/;
+  return undef unless $self->{subcommands}{$name};
+  $self->{subcommand} = $name;
+  {
+    no warnings 'redefine';
+    local *Applify::app = sub {
+      warn "Avoided deep recursion. Do not call app {} from subcommand block!\n";
+    };
```
Comment (jhthorsen) [inline lib/Applify.pm]: When does this happen? Is it just to help the author of the application? If so, do this instead:

```
local *Applify::app = sub { Carp::confess("Looks like you have a typo in your script! Cannot have app{} inside a subcommand options block.") };
```

File: lib/Applify.pm (hunk 6 above)
Comment (kiwiroy) [reply lib/Applify.pm]: TIL - really, does that work?

File: lib/Applify.pm (hunk 8 above)
Comment (kiwiroy) [reply lib/Applify.pm]: Hmm, yeah I had thoughts of removing the `$self->{subcommand}` assignment and doing a `return map { $_->{name} } grep { $_->{active} } values %{$self->{subcommands}}` in `sub subcommand {}`

File: lib/Applify.pm (hunk 10 above)
Comment (kiwiroy) [reply lib/Applify.pm]: Pretty much. `confess` `die`s that ok?

Comment (jhthorsen) [reply lib/Applify.pm]: Could be useful with confess, so you can see in which subcommand definition that you have the bug.

File: lib/Applify.pm (hunk 9 above)
Comment (kiwiroy) [reply lib/Applify.pm]: "adaptation"?

File: lib/Applify.pm (hunk 6 above)
Comment (jhthorsen) [reply lib/Applify.pm]: Yeah 😃

File: lib/Applify.pm (hunk 8 above)
Comment (jhthorsen) [reply lib/Applify.pm]: I like your use of `$self->{subcommand}`

File: lib/Applify.pm (hunk 7 above)
Comment (kiwiroy) [reply lib/Applify.pm]: trimmed the for loop - works with `@options` too.

```perl
($width) = sort { $b <=> $a } map { length $_->{name} } @options;
```

**Comment by jhthorsen** [comment]
Awesome! 0.14 is soon on CPAN 👍

**Pull request merged**